autowarefoundation/autoware.universe
# feat(default_ad_api): add motion api

## Description

Add a hook for when the vehicle starts. See related links for details.

## Related links

https://github.com/autowarefoundation/autoware/discussions/2464#discussioncomment-3698877

## Tests performed

Test with the simple planning simulator. This feature is off by default.

1. Check that the vehicle can start without calling accept start service.

Then, edit [this parameter](https://github.com/autowarefoundation/autoware.universe/blob/5d0b54de8bfb92b7451294bba9a1768ef6a611f4/system/default_ad_api/launch/default_ad_api.launch.py#L35) to True and check the following steps.

1. Check that the state is STOPPED before engagement.
1. Check that the state is STARTING and the vehicle holds a stop after engagement.
1. Check that the state is MOVING and the vehicle starts after calling accept start service.
1. Stop the vehicle for more than 1 second by an obstacle (use rviz tool).
1. Check that the state is STOPPED.
1. Remove the obstacle
1. Check that the state is STARTING and the vehicle holds a stop.
1. Check that the state is MOVING and the vehicle starts after calling accept start service.
1. Stop the vehicle for less than 1 second by an obstacle (use rviz tool).
1. Remove the obstacle
1. Check that the state is MOVING and the vehicle restarts

## Notes for reviewers

Command to check state.
```
ros2 topic echo /api/motion/state
```

Command to accept start.
```
ros2 service call /api/motion/accept_start autoware_adapi_v1_msgs/srv/AcceptStart 
```

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.


## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.


## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.


After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/
